File: algorithm/JavaScript/AllPathsFromSourceToTarget.js
/**
 * @param {number[][]} graph
 * @return {number[][]}
 */
var allPathsSourceTarget = function(graph) {
  const len = graph.length;
  const visited = new Array(len).fill(0);
  let res = [];
  const dfs = (start, end, visited, path, graph) => {
    if (start === end) {
      res.push(path.slice())
      return
    }
    for (const vertex of graph[start]) {
      if (!visited[vertex]) {
        visited[vertex] = 1
        path.push(vertex)
        dfs(vertex, end, visited, path, graph)
        path.pop()
        visited[vertex] = 0
      }
    }
  }
  dfs(0, len - 1, visited, [0], graph);
  return res;
};

File: algorithm/python/AllPathsFromSourceToTarget.py
class Solution:
    def allPathsSourceTarget(self, graph):
        visited = [0 for _ in range(len(graph))]
        self.res = []
        self.dfs(0, len(graph)-1, visited, [0], graph)
        return self.res
        
    def dfs(self, start, end, visited, path, graph):
        if start == end:
            self.res.append(path)
            return
        
        for vertex in graph[start]:
            if not visited[vertex]:
                visited[vertex] = 1
                self.dfs(vertex, end, visited, path+[vertex], graph)
                visited[vertex] = 0